ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util.hs

```diff
@@ -374,3 +388,51 @@ fib n = round $ phi ** fromIntegral n / sq5
 eitherToMaybe :: Either a b -> Maybe b
 eitherToMaybe (Left _)  = Nothing
 eitherToMaybe (Right x) = Just x
+
+{-------------------------------------------------------------------------------
+  Fuse code, i.e. regions of code that will throw an exception if accessed
+  concurrently.
+-------------------------------------------------------------------------------}
+
+-- | A WithIntactFuse expression is one of which a single instance can be
```

Comment:
I was going to suggest:

> All of these `lift`s are annoying, would a simple `withFuse :: Fuse m -> m a -> m a` combinator suffice?

Technically, it would, so `Fuse m -> ...` could replace `WithIntactFuse (...)`. However, I think without the newtype wrapper it'd be far too easy to make the mistake in which recursive calls also try to claim the fuse :/

It might be worth adding a comment explaining that this "obvious" tempting simplification was intentionally dismissed.